feat: automate segmentation tracking

stulacy opened the pull request:
Adds the ability to segment and track cells directly rather than having to run CellPose and TrackMate manually outside of CellPhe.

The outstanding features are:

- [x] ROIs extracted. The ROIManager can't be accessed headlessly, so the ROIs will need to be extracted from the XML file instead
- [x] Allow user to configure tracking method & parameters (currently hardcoded to Simple LAP and the default parameters)
- [x] Streamline CSV export by piping the XML output straight to text rather than to an intermediate file
- [x] General code refactor as the function is too big and does too many things
- [x] Add testing comparing the exported features to those exported manually

@llwiggins Here is the automated segmentation & tracking functionality. After cloning the repo and checking out this branch (`feature/automate-segmentation-tracking`) and then installing it (`pip install -e .`), it can be run with:

```
from cellphe.input import segment_images, track_images

segment_images("/path/to/directory/containing/images", "/path/to/directory/to/save/masks")
track_images("/path/to/directory/to/save/masks, "/path/to/file/for/output/feature.csv")
```

stulacy commented:
Unfortunately that error message doesn't provide that much information. What does it say in `/Users/laura/CellPhePy/hs_err_pid71175.log`?
The only other differences are the Python version (I'm on Python 12, you appear to be on 10) and a very minor difference in Java version (I'm on patch 422, you're on 402, so that really shouldn't impact much).

llwiggins commented:
Hmm upgraded to python 12 but still get the same error, .log file attached!

On Wed, 14 Aug 2024 at 14:34, Stuart Lacy ***@***.***> wrote:

> Unfortunately that error message doesn't provide that much information.
> What does it say in /Users/laura/CellPhePy/hs_err_pid71175.log?
> The only other differences are the Python version (I'm on Python 12, you
> appear to be on 10) and a very minor difference in Java version (I'm on
> patch 422, you're on 402, so that really shouldn't impact much).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uoy-research/CellPhePy/pull/65#issuecomment-2288765518>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV2ODAKZQTBXSMMXDTSIVGTZRNME7AVCNFSM6AAAAABMQGC4C6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOBYG43DKNJRHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Dr. Laura Wiggins (she/her)
Post Doctoral Research Associate

Pyne Lab
Department of Materials Science and Engineering
University of Sheffield
Sir Robert Hadfield Building
Mappin Street
Sheffield, S1 3JD

stulacy commented:
Sorry @llwiggins , to upload attachments you'll need to do so from the GitHub conversation rather than via email.

llwiggins commented:
[hs_err_pid75160.log](https://github.com/user-attachments/files/16614447/hs_err_pid75160.log)
Oops, sorry - here's the log file!

stulacy commented:
So this seems like one of those fun Java bugs that are really hard to track down! I'm going to try and track down a Mac to test it on myself, but in the meantime would you be able to test it in a virtual environment outside of conda?
It shouldn't make a huge difference, but conda does add some extra complexity.

I.e. 
`python -m venv ~/venvs/cellphepy` (create a virtual env)
`source ~/venvs/cellphepy/bin/activate` (activate it)
`cd /path/to/this/repo`
`git switch feature/automate-segmentation-tracking` (confirm we are on this branch)
`python -m pip install -e .` (install package)

Then run your script.

llwiggins commented:
Non-conda environment has done the trick, thank you Stuart! ⭐

On Wed, 14 Aug 2024 at 16:23, Stuart Lacy ***@***.***> wrote:

> So this seems like one of those fun Java bugs that are really hard to
> track down! I'm going to try and track down a Mac to test it on myself, but
> in the meantime would you be able to test it in a virtual environment
> outside of conda?
> It shouldn't make a huge difference, but conda does add some extra
> complexity.
>
> I.e.
> python -m venv ~/venvs/cellphepy (create a virtual env)
> source ~/venvs/cellphepy/bin/activate (activate it)
> cd /path/to/this/repo
> git switch feature/automate-segmentation-tracking (confirm we are on this
> branch)
> python -m pip install -e . (install package)
>
> Then run your script.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/uoy-research/CellPhePy/pull/65#issuecomment-2289104421>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AV2ODAKJMC4JQQMZ4AMEFUTZRNY7RAVCNFSM6AAAAABMQGC4C6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEOBZGEYDINBSGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Dr. Laura Wiggins (she/her)
Post Doctoral Research Associate

Pyne Lab
Department of Materials Science and Engineering
University of Sheffield
Sir Robert Hadfield Building
Mappin Street
Sheffield, S1 3JD

stulacy commented:
Phew! Conda can often hinder more than it can help, I'd definitely recommend using virtual envs over it in general.

I'll update this PR when I add new features.

stulacy commented:
Hi @llwiggins , I've got the ROIs exporting. If you pull the changes and try again, this time the `track_images` function accepts a third argument: the name of a folder to export the ROIs to. It seems to work from my perspective, but it'd be useful for someone who actually knows what they're doing with ImageJ to confirm!

I haven't added the interpolation in yet - as mentioned previously this is slightly awkward as this is currently done through a macro to the ROIManager, which doesn't work in the headless mode. I'll explore this later after I've done some tidying up and we've confirmed the ROIs are as expected.

stulacy commented:
The PR is ready for merging. The last few commits have added the following functionality:

  - Save ROIs as a zip archive, to allow them to be easily imported into ImageJ
  - Save ROIs with a consistent naming scheme. Here it is `<frameid>-<cellid>.roi`, where both `<frameid>` and `<cellid`> are 0 padded so in ImageJ they are sorted first by frame and then by cell
  - ROIs have their frame number embedded so that in ImageJ selecting an ROI opens the corresponding frame in the image viewer
  - The output CSV from `track_images` can now be read into the main CellPhe pipeline through `import_data`
  - Updated README with a guide to using the tracking and segmenting functionality
  - Scaffolding is in place for users to choose custom trackers (currently only Simple LAP is supported)
  - Fixed bugs with the original ROI exporting, namely incorrect parsing of the XML file, and extrapolating a fully enclosed boundary from the sparse points generated by TrackMate.
